kani-compiler/src/codegen_cprover_gotoc/codegen/place.rs

```diff
@@ -498,33 +509,44 @@ impl<'tcx> GotocCtx<'tcx> {
             ProjectionElem::Downcast(_, idx) => {
                 // downcast converts a variable of an enum type to one of its discriminated cases
                 let t = before.mir_typ();
-                match t.kind() {
+                let (case_name, type_or_variant) = match t.kind() {
                     ty::Adt(def, _) => {
-                        let variant = def.variants().get(idx).unwrap();
-                        let case_name = variant.name.to_string();
-                        let typ = TypeOrVariant::Variant(variant);
-                        let expr = match &self.layout_of(t).variants {
-                            Variants::Single { .. } => before.goto_expr,
-                            Variants::Multiple { tag_encoding, .. } => match tag_encoding {
-                                TagEncoding::Direct => before
-                                    .goto_expr
-                                    .member("cases", &self.symbol_table)
-                                    .member(&case_name, &self.symbol_table),
-                                TagEncoding::Niche { .. } => {
-                                    before.goto_expr.member(&case_name, &self.symbol_table)
-                                }
-                            },
-                        };
-                        ProjectedPlace::try_new(
-                            expr,
-                            typ,
-                            before.fat_ptr_goto_expr,
-                            before.fat_ptr_mir_typ,
-                            self,
-                        )
+                        let variant = def.variant(idx);
+                        (variant.name.as_str().into(), TypeOrVariant::Variant(variant))
                     }
-                    _ => unreachable!("it's a bug to reach here!"),
-                }
+                    ty::Generator(..) => {
+                        (self.generator_variant_name(idx), TypeOrVariant::GeneratorVariant(idx))
+                    }
+                    _ => unreachable!("it's a bug to reach here! {:?}", &t.kind()),
+                };
+                let layout = self.layout_of(t);
+                let expr = match &layout.variants {
+                    Variants::Single { .. } => {
+                        assert!(!t.is_generator());
```

Comment:
Might be clearer to lift these to a precondition
```rust
if t.is_generator {
    assert!(matches!(layout.variants, Variants::Single { .. });
    etc ...
}
```

Reply:
Created a function `check_generator_layout_assumptions` because this code was repeated in a few places.